Change file scoping of test helper functions

sbutcher-arm opened the pull request:
## Description
Dependent on configured options, not all of the helper functions were being used, which was leading to warning of unused functions with Clang.

To avoid any complex compile time options, or adding more logic to `generate_test_code.py` to screen out unused functions, those functions which were provoking the warning were changed to remove static, remove them from file scope, and expose them to the linker.

Fixes #2051.

## Status
**READY**

## Requires Backporting
NO  

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Changelog updated
- [x] Backported

sbutcher-arm commented:
Open question to the reviewers (and everyone else) - given this only occurs on clang, and is reproduced easily on one of our standard test scripts, `test-ref-configs.pl`, should we have an additional test in `all.sh` to test this with clang?

I'm not convinced it's worth it, but I'd like to know what you think.

gilles-peskine-arm commented:
This should be backported when #930 is backported (which I hope is soon). Having different test infrastructure in the different maintained branches causes extra work on each backport.

gilles-peskine-arm commented:
I approve of this change as a quick fix for CI, but I don't agree with the general principle that tests should compile with `-Wunused`. It's perfectly ok if we break the property that test code doesn't compile with `-Wunused` as long as our build scripts sets the proper build options, which at least the makefile does.

Patater commented:
In general, I like using `-Wunused-function` because it catches issues when I've added a function but forgot to call it. Perhaps I added a new function with a similar name and signature as another function, but had a typo and used only one of those functions on accident. If we have a solution where we can keep using `-Wunused`, I'm happier.

AndrzejKurek commented:
I would also opt for @mpg 's solution in https://github.com/ARMmbed/mbedtls/issues/1628.

sbutcher-arm commented:
Regardless of @gilles-peskine-arm's point about not needing these warnings for test code, we didn't previously need the warnings turned off, and compiled cleanly without them. This PR fixes a regression in #930.

I think the inconsistency of having the warnings disabled and but not needing them is more of a problem, than in now needing the warnings to be turned off, because we've introduced an issue. By accepting this PR, we restore the code back to the state where it was before #930.

I don't like the idea of changing the order to `$(CFLAGS) $(LOCAL_CFLAGS)` because users will expect `CFLAGS` to override the flags, not the local Makefile to override theirs. Rather than do that, we should change `test-ref-configs.pl` to remove the warnings, if we really wanted to remove the warnings.

But - I don't see a compelling reason to remove the warnings because they were enabled before, and only a regression has introduced the problem. Therefore I think it makes more sense to actively enable them in `tests/Makefile` now we've fixed them.

Patater commented on tests/Makefile:
```diff
@@ -3,7 +3,7 @@
 
 CFLAGS	?= -O2
-WARNING_CFLAGS ?= -Wall -W -Wdeclaration-after-statement -Wno-unused-function -Wno-unused-value
+WARNING_CFLAGS ?= -Wall -W -Wdeclaration-after-statement -Wunused-function -Wunused-value
```
Consider using `-Wunused` (which is -Wunused-function -Wunused-label -Wunused-parameter -Wunused-value -Wunused-variable), if we can.

gilles-peskine-arm replied:
Remove `-Wunused-function -Wunused-value`. They're already included in `-Wall`.

gilles-peskine-arm commented:
I'm actually slightly surprised that it was so easy to avoir unused functions — or did you only test the reference configurations? Some test suites define auxiliary functions and I would have thought that some of those aren't used in all configurations. To add accurate `#ifdef` guards around all of them would not have any benefit other than silencing the compiler.

gilles-peskine-arm commented:
We test a bunch of other configurations in `all.sh`, including `depends-*.pl`. I wouldn't be surprised if this broke `depends-pkalgs.pl` because of some auxiliary function that's only used for RSA, for example. This PR needs an `all.sh` run before gatekeeping.

mpg commented:
@sbutcher-arm 

> we didn't previously need the warnings turned off, and compiled cleanly without them.

I'm not sure what you mean by that, but the test Makefile has been turning off these warnings since forever (ie at least polarssl-1.0.0) so I think we _did_ previously need the warnings turned off. We were not compiling cleanly without them, we were compiling cleanly _because_ they were active by default.

```
% git show polarssl-1.0.0:tests/Makefile | grep unused
                        -Wno-unused-function -Wno-unused-value
```

> This PR fixes a regression in #930.

I don't think that's the case. #1628 happens with 2.9.0, ie well before #930 was merged, and I just checked that I can reproduce #2051 with 2.9.0 as well. I think #930 is a red herring here, and the real culprit is elsewhere.

mpg commented:
Note: our good friend git bisect run thinks `CC=clang tests/scripts/test-ref-configs.pl config-suite-b.h` started to fail in f95c1764faa8218c95b1f03b4d0cf7a1ad60f1e2. For reference, the first release containing this commit is 2.4.2. Direct testing confirms `CC=clang tests/scripts/test-ref-configs.pl config-suite-b.h` also fails in the 2.7 branch (but not in the 2.1 branch). I'm therefore applying the "need backports" label.

As @gilles-peskine-arm   noted, the issue is caused by using `-Wall -Wextra` in `CFLAGS`, but I fully agree with @sbutcher-arm that this should be supported because it's such an intuitive thing to do. This doesn't cause any issue with GCC however because with GCC `-Wno-xxx -Wall -Wextra` still disables xxx, but with Clang it enables warning xxx (assuming it's included in all or extra).

I'm still inclined to change the order of the flags so that the Makefiles behave (1) consistently between GCC and Clang and (2) as they were intended to since basically forever.

However, I've noted @sbutcher-arm 's point that overriding CFLAGS might surprise some users (even though that's what has been happening with GCC for a very long time) and @Patater 's point that we might want to have the ununsed warnings in test scripts too after all. So I could also accept the PR as it stands, as long as we agree that this is a deviation from our historical behaviour (which is absolutely fine if that's what we want) and _not_ just reverting a bug supposedly introduced by #930.

sbutcher-arm commented:
Backported to 2.7 with PR #2320.

AndrzejKurek commented:
@Patater - please re-review this PR, as it is targeted at 19.02.

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -1,5 +1,10 @@
+= mbed TLS x.x.x branch released xxxx-xx-xx
+
+Bugfix
+   * Fix to allow test-ref-config.pl to build with Clang. Fixes #1628.
```
There's something wrong with the changelog entry: #1628 says nothing about `test-ref-configs.pl`.
```suggestion
   * Allow building test suites with "clang -Wall -Wextra -Wunused -Werror". Fixes #1628.
```

Patater commented on tests/Makefile:
```diff
@@ -3,7 +3,7 @@
 
 CFLAGS	?= -O2
-WARNING_CFLAGS ?= -Wall -W -Wdeclaration-after-statement -Wno-unused-function -Wno-unused-value
+WARNING_CFLAGS ?= -Wall -W -Wdeclaration-after-statement -Wunused
```
Minor: `-Wall` will also set many `-Wunused-*` options. I don't think we need to set `-Wunused` explicitly, but it doesn't hurt to set it here.

sbutcher-arm commented:
I've changed the ChangeLog entry. Note, I haven't rebased as then I'd need to create a new backport PR for `mbedtls-2.16`.

@mpg / @gilles-peskine-arm - Can you please re-review?

gilles-peskine-arm commented:
Applying this change to Mbed Crypto would break the build of some PSA crypto test suites. I think the change is ok for Mbed TLS as it is now, but it would be better to get CI feedback first, so I made a branch with these commits cherry-picked to development and started a CI run: https://jenkins-internal.mbed.com/job/mbedtls-release-new/281/

If this CI passes then this PR should be merged to both `mbedtls-2.16` and `development`.

I'll make a PR for Mbed Crypto so that we can merge this without breaking CI.
